File: src/PHP/scale.php
<?php

use PHP2xAI\Runtime\PHP\Core\GraphRuntime;
use PHP2xAI\Tensor\Tensor;

include("../../vendor/autoload.php");

function runScale(string $name, Tensor $input, float $scale): void
{
	$input->setRequiresGrad(true);
	$output = $input->scale($scale);
	$runtime = GraphRuntime::createFromOutputTensor($output);
	$runtime->forward();
	$runtime->backward();
	$runtime->refreshTensorsData();

	echo "\n=== {$name} ===\n";
	echo "scale: {$scale}\n";
	$output->printData();
	$input->printGrad();
}

runScale('Positive scalar on a matrix', Tensor::createFromData([
	[1.0, -2.0],
	[3.0, -4.0],
], 'matrix'), 2.5);

runScale('Negative scalar on token embeddings', Tensor::createFromData([
	[[1.0, 2.0], [3.0, 4.0]],
	[[5.0, 6.0], [7.0, 8.0]],
], 'tokens'), -0.5);

runScale('Zero scalar', Tensor::createFromData([1.0, -2.0, 3.0], 'vector'), 0.0);

File: src/Torch/scale.py
import torch


def run_scale(name: str, values: list, scale: float) -> None:
    input_tensor = torch.tensor(values, dtype=torch.float32, requires_grad=True)
    output = input_tensor * scale
    output.backward(torch.ones_like(output))

    print(f"\n=== {name} ===")
    print("scale:", scale)
    print("output:")
    print(output)
    print("input grad:")
    print(input_tensor.grad)


run_scale("Positive scalar on a matrix", [[1.0, -2.0], [3.0, -4.0]], 2.5)
run_scale(
    "Negative scalar on token embeddings",
    [[[1.0, 2.0], [3.0, 4.0]], [[5.0, 6.0], [7.0, 8.0]]],
    -0.5,
)
run_scale("Zero scalar", [1.0, -2.0, 3.0], 0.0)
